File: LeetCode/Arrays/896-Monotonic-Array.js
/**
 * Problem: 896. Monotonic Array
 * Platform: LeetCode
 * Difficulty: Easy
 *
 * Approach:
 * Track whether the array can still be:
 * 1. Non-decreasing
 * 2. Non-increasing
 *
 * If nums[i] > nums[i - 1], it cannot be non-increasing.
 * If nums[i] < nums[i - 1], it cannot be non-decreasing.
 *
 * If either condition remains true, the array is monotonic.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums
 * @return {boolean}
 */
var isMonotonic = function(nums) {
    let isIncreasing = true;
    let isDecreasing = true;

    for (let i = 1; i < nums.length; i++) {
        if (nums[i] > nums[i - 1]) {
            isDecreasing = false;
        }

        if (nums[i] < nums[i - 1]) {
            isIncreasing = false;
        }
    }

    return isIncreasing || isDecreasing;
};

File: LeetCode/Arrays/896-Monotonic-Array.py
"""
Problem: 896. Monotonic Array
Platform: LeetCode
Difficulty: Easy

Approach:
Track whether the array can still be:
1. Non-decreasing
2. Non-increasing

If we ever find:
- nums[i] > nums[i-1], it cannot be non-increasing.
- nums[i] < nums[i-1], it cannot be non-decreasing.

If either condition remains true, the array is monotonic.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List

class Solution:
    def isMonotonic(self, nums: List[int]) -> bool:
        is_increasing = True
        is_decreasing = True

        for i in range(1, len(nums)):
            if nums[i] > nums[i - 1]:
                is_decreasing = False

            if nums[i] < nums[i - 1]:
                is_increasing = False

        return is_increasing or is_decreasing